src/actions/languages/python/motion.ts

```diff
@@ -0,0 +1,232 @@
+    const lines = [...this.lines(document)]; // convert generator to Array
+    const infos = lines.map((text, index) => this._parseLine(index, text));
+
+    return infos.filter((x) => x) as LineInfo[]; // filter out empty/comment lines (undefined info)
+  }
+
+  static _parseStructure(lines: LineInfo[]): StructureElement[] {
+    const last = lines.length;
+    const structure: StructureElement[] = [];
+
+    for (let index = 0; index < last; index++) {
+      const info = lines[index];
+      const text = info.text;
+
+      if (text.match(PythonDocument.reDefOrClass)) {
+        const type = text.match(PythonDocument.reDef) ? 'function' : 'class';
+
+        // Find the end of the current function/class
+        let idx = index + 1;
+
+        for (; idx < last; idx++) {
+          if (lines[idx].indentation <= info.indentation) {
+            break;
+          }
+        }
+
+        // Since we stop when we find the first line with a less indentation
+        // we pull back one line to get to the end of the function/class
+        idx--;
+
+        const endLine = lines[idx];
+
+        structure.push({
+          type,
+          start: new Position(info.line, info.indentation),
+          // Calculate position of last non-white character)
+          end: new Position(
+            endLine.line,
+            endLine.text.search(PythonDocument.reLastNonWhiteSpaceCharacter) - 1
+          ),
+        });
+      }
+    }
+
+    return structure;
+  }
+
+  /*
+   * Find the position of the specified:
+   *    type: function or class
+   *    direction: next or prev
+   *    edge: start or end
+   *
+   * With this information one can determine all of the required motions
+   */
+  find(type: Type, direction: Direction, edge: Edge, position: Position): Position | undefined {
+    // Choose the ordering method name based on direction
+    const isDirection = direction === 'next' ? 'isAfter' : 'isBefore';
+
+    // Filter function for all elements whose "edge" is in the correct "direction"
+    // relative to the cursor's position
+    const dir = (element: StructureElement) => element[edge][isDirection](position);
+
+    // Filter out elements from structure based on type and direction
+    const elements = this.structure.filter((elem) => elem.type === type).filter(dir);
+
+    if (edge === 'end') {
+      // When moving to an 'end' the elements should be started by the end position
+      elements.sort((a, b) => a.end.line - b.end.line);
+    }
+
+    // Return the first match if any exist
+    if (elements.length) {
+      // If direction === 'next' return the first element
+      // otherwise return the last element
+      const index = direction === 'next' ? 0 : elements.length - 1;
+      const element = elements[index];
+      const pos = element[edge];
+
+      // execAction MUST return a fully realized Position object created using new
+      return new Position(pos.line, pos.character);
```

Comment:
I think `return pos;` is fine, actually - `Position`s are immutable

Reply:
Done